sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -265,14 +265,21 @@ abstract class Expression extends TreeNode[Expression] {
     case single => single :: Nil
   }
 
+  protected def flatArgumentStrings: Iterator[String] = flatArguments.map {
+    case e: Expression => e.argumentString
+    case arg => s"$arg"
+  }
+
   // Marks this as final, Expression.verboseString should never be called, and thus shouldn't be
   // overridden by concrete classes.
   final override def verboseString(maxFields: Int): String = simpleString(maxFields)
 
   override def simpleString(maxFields: Int): String = toString
 
   override def toString: String = prettyName + truncatedString(
-    flatArguments.toSeq, "(", ", ", ")", SQLConf.get.maxToStringFields)
+    flatArgumentStrings.toSeq, "(", ", ", ")", SQLConf.get.maxToStringFields)
+
+  def argumentString: String = toString
```

Comment:
is it possible to only add one method?

Reply:
IMO, the `argumentString` is needed because `AttributeReference` already overwrite `toString`, thus we need the new abstract string function to switch to non-exprid format.  For `flatArgumentStrings`, it only have two callers. I refactored the `toAggString` of `AggregateFunction`, then we don't need to add the method `flatArgumentStrings` in `Expression` but just implement it within `toString`. See commit 51bb46aea0ae7a4c74c6e1026b173cafbfd2ca99. @cloud-fan